### ISSUE TITLE
Windows-style path is not recognized under cygwin

### ISSUE THREAD

**Issue by 05f74a29-ec3f-4076-aea4-26856f01b326** [python/cpython#90907]
BPO | [46751](https://bugs.python.org/issue46751)
--- | :---
Nosy | @pfmoore, @ericvsmith, @tjguk, @zware, @eryksun, @zooba, @lazka, @mikekaganski

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-02-14.18:42:01.267>
labels = ['type-bug', '3.9', 'OS-windows']
title = 'Windows-style path is not recognized under cygwin'
updated_at = <Date 2022-02-15.07:40:35.403>
user = 'https://github.com/mikekaganski'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-15.07:40:35.403>
actor = 'lazka'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Windows']
creation = <Date 2022-02-14.18:42:01.267>
creator = 'mikekaganski'
dependencies = []
files = []
hgrepos = []
issue_num = 46751
keywords = []
message_count = 8.0
messages = ['413247', '413250', '413256', '413259', '413261', '413264', '413272', '413277']
nosy_count = 8.0
nosy_names = ['paul.moore', 'eric.smith', 'tim.golden', 'zach.ware', 'eryksun', 'steve.dower', 'lazka', 'mikekaganski']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue46751'
versions = ['Python 3.9']
```

</p></details>

**Comment by Sporking** [comment]
I disagree that this is "expected behavior". The majority of Cygwin programs (including languages such as Perl, etc.) work just fine with Windows paths. This is by design, to avoid Windows users having to constantly call 'cygpath' every time they execute something from the command line, or want to pass it a path pasted from a Windows Explorer window or other source that produces Windows paths. See here for some documentation of this:

https://cygwin.com/cygwin-ug-net/using.html#pathnames-win32

I don't know to what degree this support for Windows paths is due to the Python distribution itself, versus to some kind of patching that the Cygwin team or the Cygwin support libraries may do to it to add support for Windows, but it generally works, at least for paths that don't have Windows-specific things like embedded spaces.

Up until recently, this worked fine for Python as well. In particular, Cygwin's version of Python 3.8 worked fine in this regard. When we upgraded to Python 3.9, things suddenly broke, and we saw the behavior described in this bug report. I think that this behavior is a regression between versions 3.8 and 3.9, and I would very much like it to be fixed. I'm not sure who is responsible for fixing it, however.

Some evidence that this is a regression:
`
C:\example>copy con test.py
#!/usr/bin/env python3
print('hello')
^Z

C:\example>c:\cygwin\bin\python3.8 test.py
hello

C:\example>c:\cygwin\bin\python3.9 test.py
hello

C:\example>c:\cygwin\bin\python3.8 C:\example\test.py
hello

C:\example>c:\cygwin\bin\python3.9 C:\example\test.py
/usr/bin/python3.9: can't open file '/cygdrive/c/example/C:\example\test.py': [Errno 2] No such file or directory
`

Cygwin Python 3.9 apparently does not recognize that paths starting with "C:\" are absolute Windows paths, and so it tries to join them to the current directory as if they were relative paths before trying to invoke the python script.

Any advice on how to get a bug report for this regression to the right team as soon as reasonably possible? This bug is very inconvenient, and breaks a *lot* of our scripts. It makes Python dramatically less useable than other languages (Perl, Ruby, etc.) under Cygwin, and may force us to downgrade to Python 3.8 until it is fixed.

**Comment by pfmoore** [comment]
> Any advice on how to get a bug report for this regression to the right team as soon as reasonably possible?

You should report it to the cygwin python maintainers (whoever they are, I'm afraid I don't know). Either they will be able to fix it (most likely) or they will be able to pinpoint what core Python behaviour change is affecting this.

**Comment by joro88** [comment]
I absolutely agree with @Sporking

All other cygwin programs take windows paths and work fine.

The workaround what @pfmoore suggested makes usage of cygwin python pretty hard. Especially when some bat files should call python. If they call cygpath and current installation of python is native, it would fail. There are at least 3 cases - native, cygwin, mingw. Installation via package managers could introduce more cases. So such workaround will be pretty dirty.

Bat files are immature and such checks are not so easy to implement, read and support.

**I recommend never to use python via cygwin.** Install native package. If 2.5 years there is no progress, I expect nothing good for nearby future. It's a timewaster.

**Comment by ericvsmith** [comment]
@joro88 : Agreed, unfortunately. I've basically given up on Cygwin python, and use the native Windows version instead.

**Comment by gavenkoa** [comment]
I'm working with Mercurial. It cannot use Windows style paths because internally it resorts to `os.path.realpath`, which is broken:

```
(Pdb) p os.path.realpath('c:/Users')
'/cygdrive/e/my/c:/Users'
```

As you see `realpath` prepend current directory. It basically breaks everything.

By interactive debug session, demonstrating the problem:

```
>>> from mercurial import hg, ui
>>> UI = ui.ui()
>>> import pdb
>>> pdb.run("hg.repository(UI, path=b'c:/Users/user/my/utils', create=False)")
...
(Pdb) w
  /usr/lib/python3.9/bdb.py(580)run()
-> exec(cmd, globals, locals)
  <string>(1)<module>()
  /usr/lib/python3.9/site-packages/mercurial/hg.py(240)repository()
-> repo = cls.instance(
  /usr/lib/python3.9/site-packages/mercurial/hg.py(153)instance()
-> return cls.instance(ui, path, *args, **kwargs)
  /usr/lib/python3.9/site-packages/mercurial/localrepo.py(3625)instance()
-> repo = repo_maker()
  /usr/lib/python3.9/site-packages/mercurial/localrepo.py(3623)repo_maker()
-> return makelocalrepository(ui, localpath, intents=intents)
(Pdb) p hgpath
b'/cygdrive/e/my/c:/Users/user/my/utils/.hg'
```

**Comment by pfmoore** [comment]
As noted above, if you want to see progress on this, you'll need to report it to the maintainers of the cygwin Python package in the first instance.

I'll also note that at this point, Python 3.9 is long out of bugfix support (and nearly out of security support) and you should upgrade to a newer version of Python.
